core/ts/MaciState.ts

```diff
@@ -717,6 +728,169 @@ class Poll {
+            this.cbi = 0
+        } 
+        return
+    }
+
+    public previousSubsidyIndexToString = ():string => {
+        const batchSize = this.batchSizes.subsidyBatchSize
+        const numBatches = Math.ceil(this.ballots.length/batchSize)
+        let cbi = this.cbi
+        let rbi = this.rbi
+        if (this.cbi === 0 && this.rbi === 0) {
+            return "0-0"
+        }
+        if (this.cbi !== 0 ) {
+            cbi --
+        } else {
+            rbi --
+            cbi = numBatches - 1
+        }
+        return rbi.toString() + "-" + cbi.toString()
+    }
+
+    public coefficientCalculation = (rowBallot: Ballot, colBallot: Ballot): BigInt  => {
+        let sum = BigInt(0)
+        for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+            sum += BigInt(rowBallot.votes[p]) * BigInt(colBallot.votes[p])
+        }
+        let res = BigInt(this.MM * (10 ** this.WW))/(BigInt(this.MM)+BigInt(sum))
+        return res
+    }
+
+    public subsidyCalculation = (rowStartIndex:number, colStartIndex:number): Ballot[][] => {
+        const batchSize = this.batchSizes.subsidyBatchSize
+        let ballots1: Ballot[] = [] 
+        let ballots2: Ballot[] = [] 
+        const emptyBallot = new Ballot(
+            this.maxValues.maxVoteOptions,
+            this.treeDepths.voteOptionTreeDepth,
+        )
+        for (let i = 0; i < batchSize; i++) {
+            const row = rowStartIndex + i
+            const col = colStartIndex + i
+            const rowBallot = (row < this.ballots.length)?this.ballots[row]:emptyBallot
+            const colBallot = (col < this.ballots.length)?this.ballots[col]:emptyBallot
+            ballots1.push(rowBallot)
+            ballots2.push(colBallot)
+        }
+        for (let i = 0; i < batchSize; i++) {
+            for (let j = 0; j < batchSize; j++) {
+                const row = rowStartIndex + i
+                const col = colStartIndex + j
+                const rowBallot = (row < this.ballots.length)?this.ballots[row]:emptyBallot
+                const colBallot = (col < this.ballots.length)?this.ballots[col]:emptyBallot
+
+                const kij = this.coefficientCalculation(rowBallot, colBallot)
+                for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+                    const vip = BigInt(rowBallot.votes[p])
+                    const vjp = BigInt(colBallot.votes[p])
+                    this.subsidy[p] = BigInt(this.subsidy[p]) + BigInt(kij) * vip * vjp
+                }
+            }
+        }
+
+        if (rowStartIndex === colStartIndex) {
+            for (let i = 0; i < batchSize; i++) {
+                const idx = rowStartIndex + i
+                const diagBallot = (idx < this.ballots.length)?this.ballots[idx]:emptyBallot
+                const kii = this.coefficientCalculation(diagBallot, diagBallot)
+                for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+                    const vip = BigInt(diagBallot.votes[p])
```

Comment:
```suggestion
                    const vip = BigInt(diagBallot.votes[p].valueOf())
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)